c7n/actions/network.py

```diff
@@ -40,127 +48,184 @@ class ModifyVpcSecurityGroupsAction(Action):
     }
 
-    def get_groups(self, resources, metadata_key=None):
-        """Parse policies to get lists of security groups to attach to each resource
+    SYMBOLIC_SGS = set(('all', 'matched', 'network-location'))
+
+    sg_expr = None
+    vpc_expr = None
+
+    def validate(self):
+        sg_filter = self.manager.filter_registry.get('security-group')
+        if not sg_filter or not sg_filter.RelatedIdsExpression:
+            raise PolicyValidationError(self._format_error((
+                "policy:{policy} resource:{resource_type} does "
+                "not support {action_type} action")))
+        if self.get_action_group_names():
+            vpc_filter = self.manager.filter_registry.get('vpc')
+            if not vpc_filter or not vpc_filter.RelatedIdsExpression:
+                raise PolicyValidationError(self._format_error((
+                    "policy:{policy} resource:{resource_type} does not support "
+                    "security-group names only ids in action:{action_type}")))
+            self.vpc_expr = jmespath.compile(vpc_filter.RelatedIdsExpression)
+        if self.sg_expr is None:
+            self.sg_expr = jmespath.compile(
+                self.manager.filter_registry.get('security-group').RelatedIdsExpression)
+        if 'all' in self._get_array('remove') and not self._get_array('isolation-group'):
+            raise PolicyValidationError(self._format_error((
+                "policy:{policy} use of action:{action_type} with "
+                "remove: all requires specifying isolation-group")))
+        return self
+
+    def get_group_names(self, groups):
+        names = []
+        for g in groups:
+            if g.startswith('sg-'):
+                continue
+            elif g in self.SYMBOLIC_SGS:
+                continue
+            names.append(g)
+        return names
+
+    def get_action_group_names(self):
+        return self.get_group_names(
+            list(itertools.chain(
+                *[self._get_array('add'),
+                  self._get_array('remove'),
+                  self._get_array('isolation-group')])))
+
+    def _format_error(self, msg, **kw):
+        return msg.format(
+            policy=self.manager.ctx.policy.name,
+            resource_type=self.manager.type,
+            action_type=self.type,
+            **kw)
+
+    def _get_array(self, k):
+        v = self.data.get(k, [])
+        if isinstance(v, six.string_types):
+            return [v]
+        return v
+
+    def get_groups_by_names(self, names):
+        """Resolve security names to security groups resources."""
+        if not names:
+            return []
+        client = utils.local_session(
+            self.manager.session_factory).client('ec2')
+        sgs = self.manager.retry(
+            client.describe_security_groups,
+            Filters=[{
+                'Name': 'group-name', 'Values': names}]).get(
+                    'SecurityGroups', [])
+        if len(sgs) != len(names):
```

Comment:
Thank you for refactoring @kapilt!!

QQ, Group Names are unique per VPC, so if there are multiple vpcs in a region, would this check fail?
Example:
```
names = [ "EnterpriseRequiredSG" ]
sgs = [ 
              { "GroupName": "EnterpriseRequiredSG", "GroupId": "sg-123456", "VpcId": "vpc-123456",...},
              { "GroupName": "EnterpriseRequiredSG", "GroupId": "sg-789012", "VpcId": "vpc-789012",...} 
           ]
```
len(sgs) = 2 
len(names) = 1

Reply:
thanks, i went ahead and made this an iter through the sgs and remove sg names from a set, if set non empty at end fail.